main.go

```diff
@@ -187,66 +149,101 @@ func main() {
 	stepconf.Print(cfg)
 	fmt.Println()
 
-	if err := cfg.validateEnvs(); err != nil {
+	if err := cfg.validateArtifact(); err != nil {
 		failf("Input error: %s", err)
 	}
 
-	password := string(cfg.Password)
-	if string(cfg.AppPassword) != "" {
-		password = string(cfg.AppPassword)
+	authInputs := appleauth.Inputs{
+		Username:            cfg.ItunesConnectUser,
+		Password:            string(cfg.Password),
+		AppSpecificPassword: string(cfg.AppSpecificPassword),
+		APIIssuer:           cfg.APIIssuer,
+		APIKeyPath:          string(cfg.APIKeyPath),
 	}
-
-	filePth := cfg.IpaPath
-	if cfg.PkgPath != "" {
-		filePth = cfg.PkgPath
+	if err := authInputs.Validate(); err != nil {
+		failf("Issue with authentication related inputs: %v", err)
 	}
 
-	xcpath, err := xcodePath()
+	//
+	// Select and fetch Apple authenication source
+	authSources, err := parseAuthSources(cfg.BitriseConnection)
 	if err != nil {
-		failf("Failed to find Xcode path, error: %s", err)
+		failf("Invalid input: unexpected value for Bitrise Apple Developer Connection (%s)", cfg.BitriseConnection)
 	}
 
-	log.Printf("Xcode path: %s", xcpath)
-	fmt.Println()
+	var devportalConnectionProvider *devportalservice.BitriseClient
+	if cfg.BuildURL != "" && cfg.BuildAPIToken != "" {
+		devportalConnectionProvider = devportalservice.NewBitriseClient(http.DefaultClient, cfg.BuildURL, string(cfg.BuildAPIToken))
+	} else {
+		fmt.Println()
+		log.Warnf("Connected Apple Developer Portal Account not found. Step is not running on bitrise.io: BITRISE_BUILD_URL and BITRISE_BUILD_API_TOKEN envs are not set")
+	}
+	var conn *devportalservice.AppleDeveloperConnection
+	if cfg.BitriseConnection != "off" && devportalConnectionProvider != nil {
+		var err error
+		conn, err = devportalConnectionProvider.GetAppleDeveloperConnection()
+		if err != nil {
+			handleSessionDataError(err)
+		}
 
-	xcodeVersion, err := utility.GetXcodeVersion()
+		if conn == nil || (conn.APIKeyConnection == nil && conn.AppleIDConnection == nil) {
```

Comment:
Minor note: `conn != nil &&` or otherwise 2 warnings are printed out, one by handleSessionDataError and one here.